### PR TITLE
feat: normalize migrations file

### PR DESCRIPTION
**Pull request by BirajMainali**
- [x] I've read the guidelines for [contributing](CONTRIBUTING.md) and seen the [walkthrough](https://youtu.be/9OMxy1wal1s?t=1869)

If we use the reserve key character while adding a new migration it doesn't work as aspected for eg `dotnet ef migrations add "add: user"`, this tries to create a migration file same as provided name, but if the migrations name includes a reserve key characters for the file name, I don't know how but it's an unknown migrations file.

## Steps to reproduce
 - Adding initial migration including symbol `:` which is not allowed to include on the file name.
image
 - It creates an unknown file, some times it will be hard to remove from the machine.
image

_So, I've tried to fix this issue, I apologize if this brings the breaking changes._ 

- [x] Tests for the changes have been added (for bug fixes/features)
- [x] Code follows the same patterns and style as existing code in this repo

### CODE REVIEW

**Comment by ajcvickers** [comment]
@BirajMainali Thanks for your interest in this issue. Rather than generating an error, the invalid input should be sanitized so that it can still be used. Also, remember that we can't accept a PR without appropriate test coverage.

**Comment by BirajMainali** [comment]
@bricelam, can you please suggest to me, where should i include the sensitization test. based on my last commit.

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/efcore?pullRequest=28367) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/efcore?pullRequest=28367) before we can accept your contribution.<br/><br/>:x: BirajMainali [sign now](https://cla.dotnetfoundation.org/dotnet/efcore?pullRequest=28367)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/efcore?pullRequest=28367) it.</sub>

File: src/ef/Commands/MigrationsAddCommand.cs [hunk 1]
```diff
@@ -18,6 +20,9 @@ protected override void Validate()
             {
                 throw new CommandException(Resources.MissingArgument(_name.Name));
             }
+
+            // sanitizing migration name. If there is a reserve character in the migration name
+            _name.Name = new string(_name.Name.Where(c => !Path.GetInvalidFileNameChars().Contains(c)).ToArray());
```
[the remark this answers is not in the document]
Comment (BirajMainali) [reply src/ef/Commands/MigrationsAddCommand.cs]: @bricelam I made some changes, Please review.

**Comment by bricelam** [comment]
For tests, you can add them to [test/EFCore.Design.Tests/Migrations/Design/MigrationScaffolderTest.cs](https://github.com/dotnet/efcore/blob/main/test/EFCore.Design.Tests/Migrations/Design/MigrationScaffolderTest.cs).

The TempDirectory test helper will probably come in handy for these.
```cs
// Creates a temporary directory that gets deleted at the end of the test
using var projectDir = new TempDirectory();
scaffolder.Save(projectDir, migration, outputDir: null);
```

**Comment by BirajMainali** [comment]
@bricelam, I've changed the methodology to identify invalid file name characters. At previous changes I've used `Path.GetInvalidFileNameChars()` method to get invalid chars, but in .**net 6** this told that only `/`  is the invalid character for a file which is not enough. I've visited issue regarding provider method named `Path.GetInvalidFileNameChars()` on https://github.com/dotnet/runtime/issues/63383

I found some of the common and used invalid file name characters to sensitize the migrations file name. And it just works as expected. In case of other characters are found we can simply include them on the array.

**Comment by BirajMainali** [comment]
@ajcvickers @bricelam any updates regarding this PR ?

**Comment by BirajMainali** [comment]
@bricelam @ajcvickers any update regarding this PR

**Comment by BirajMainali** [comment]
@bricelam, should i know why this pr is still open

File: src/EFCore.Design/Migrations/Design/MigrationsScaffolder.cs [hunk 2]
```diff
@@ -402,6 +414,13 @@ public virtual MigrationFiles Save(string projectDir, ScaffoldedMigration migrat
         };
     }
 
+    private static string GetNormalizedMigrationId(string migrationId)
+    {
+        char[] invalidChars = { '#', '%', '&', '{', '}', '\\', '<', '>', '*', '?', '/', '$', '!', ':', '@', '+', '=', '"', '|' };
```
Comment (bricelam) [inline src/EFCore.Design/Migrations/Design/MigrationsScaffolder.cs:419]: I don't like this hard-coded list--it doesn't give different platforms a chance to have their say. Any reason you don't use `Path.GetInvalidFileNameChars`?

File: test/EFCore.Design.Tests/Migrations/Design/MigrationScaffolderTest.cs [hunk 3]
```diff
@@ -37,6 +37,17 @@ public void ScaffoldMigration_handles_generic_contexts()
...
     }
 
+    [Theory]
+    [InlineData("feat:test", "feat_test")]
+    public void Save_normalizes_migration_name(string actualValue, string expectedValue)
+    {
+        using var projectDir = new TempDirectory();
+        var scaffolder = CreateMigrationScaffolder<ContextWithSnapshot>();
+        var migration = scaffolder.ScaffoldMigration(actualValue, "WebApplication1");
+        var migrationFile = scaffolder.Save(projectDir, migration, outputDir: null);
```
Comment (bricelam) [inline test/EFCore.Design.Tests/Migrations/Design/MigrationScaffolderTest.cs:47]: I'd like to see a test that sets `outputDir`. I don't see any issues in the current code, but I think an earlier revision had some bugs around that.

File: src/EFCore.Design/Migrations/Design/MigrationsScaffolder.cs (hunk 2 above)
Comment (BirajMainali) [reply src/EFCore.Design/Migrations/Design/MigrationsScaffolder.cs:419]: from dotnet 6 `Path.GetInvalidFileNameChars` is not providing invalid enough characters as before, this only says only `/` is invalid chars which I've already mentioned https://github.com/dotnet/efcore/pull/28367#issuecomment-1185708929

Comment (bricelam) [reply src/EFCore.Design/Migrations/Design/MigrationsScaffolder.cs:418]: I thought they concluded that GetInvalidFileNameChars was a different API than the one you were originally using that *would* provide the expected characters.

File: src/EFCore.Design/Migrations/Design/MigrationsScaffolder.cs [hunk 4]
```diff
@@ -379,8 +389,9 @@ public virtual MigrationFiles Save(string projectDir, ScaffoldedMigration migrat
     {
         var lastMigrationFileName = migration.PreviousMigrationId + migration.FileExtension;
```
Comment (bricelam) [inline src/EFCore.Design/Migrations/Design/MigrationsScaffolder.cs:390]: We'll need to apply the same sanitization to lastMigrationFileName too so we can find the correct sibling.

**Comment by BirajMainali** [comment]
@bricelam, @ajcvickers

File: src/EFCore.Design/Migrations/Design/MigrationsScaffolder.cs (hunk 2 above)
Comment (roji) [reply src/EFCore.Design/Migrations/Design/MigrationsScaffolder.cs:421]: All these characters - except for `/` - are valid on Unix systems.

Comment (BirajMainali) [reply src/EFCore.Design/Migrations/Design/MigrationsScaffolder.cs:421]: @roji, I'm not sure if I need to modify this constant. I encountered a problem on a Windows operating system and found that these characters are commonly causing issues. However, I'm not sure if this is the case for other systems, so normalization may be a solution. Can you confirm if this constant needs to be modified or if there is a better approach to ensure compatibility across different systems?

**Comment by ajcvickers** [comment]
We discussed this again and we're concerned about platform differences making this unreliable and/or overly aggressive, especially since migrations may be used on a different platform than the one on which they were generated. The value here seems minimal, so instead of doing this people should create migration names that avoid using special characters, unless they are confident that these characters will work for their target platform(s).

**Comment by BirajMainali** [comment]
> We discussed this again and we're concerned about platform differences making this unreliable and/or overly aggressive, especially since migrations may be used on a different platform than the one on which they were generated. The value here seems minimal, so instead of doing this people should create migration names that avoid using special characters unless they are confident that these characters will work for their target platform(s).

@ajcvickers, I see this issue from a different view, In our office workspace we use a Windows machine for developing work and I and other team members use Linux at home, so in this case we are facing this issue. we cannot directly pull and start working due to this file system issue. Linux users can name the migrations like `feat: add product rate` in this case windows users cannot work with that migration file.  Also, we are not normalizing every normal migration file only in case of special cases we are normalizing special characters by `_` so this will not bring any issues.

**Pull request closed without merging**